src/users/controller/userController.ts

```diff
@@ -1,55 +1,147 @@
 import { UpdateUserDto } from "../dto/userUpdateDto";
 import { User } from "../entity/user";
+import { sendEmail } from '../../utils/sendEmail';
+import {JwtService} from '@nestjs/jwt';
+import {AuthGuard} from '../../middlewares/checkJwt'
+import * as bcrypt from 'bcrypt';
+import * as jwt from "jsonwebtoken";
+import {JwtPayload} from "jsonwebtoken";
+import {Token} from '../../middlewares/jwtDecorator';
 
 @Controller("users")
 export class UsersController {
+  constructor(
+    private jwtService: JwtService
+  ) {}
+
   @Inject()
   usersService: UsersService;
-
   @UseInterceptors(ClassSerializerInterceptor)
-  @Post()
-  create(@Body() createUserDto: CreateUserDto): Promise<User> {
+  @Post('create')
+  async create(@Body() createUserDto: CreateUserDto): Promise<User> {
+    const hashedPassword = await bcrypt.hash(createUserDto.password, 12);
+    createUserDto.password = hashedPassword;
     return this.usersService.create(createUserDto);
   }
 
+
   @UseInterceptors(ClassSerializerInterceptor)
+  @UseGuards(AuthGuard)
   @Get()
-  async findAll(): Promise<User[]> {
+  findAll(): Promise<Array<User>> {
     return this.usersService.findAll();
   }
 
   @UseInterceptors(ClassSerializerInterceptor)
-  @Get(":id")
-  async findOne(@Param("id") id: number): Promise<User> {
-    return await this.usersService.findOne(id);
+  @Post('login')
+  async login(
+    @Body() login: {email: string, password: string},
+    @Response({passthrough: true}) response
+  ): Promise<string> {
+    const user =  await this.usersService.findOneEmail(login.email);
+    if (!user) {
+      throw new BadRequestException('invalid credentials1');
+    }
+    if (!await bcrypt.compare(login.password, user.password)) {
+      throw new BadRequestException('invalid credentials2');
+    }
+    const jwt = await this.jwtService.signAsync({id: user.id},
+      {secret: process.env.JWT_ACCESS, expiresIn: '1h' });
+    return jwt ;
   }
 
   @UseInterceptors(ClassSerializerInterceptor)
-  @Put(":id")
+  @UseGuards(AuthGuard)
+  @Get('me')
+  async findOne(@Token() data: number): Promise<User> {
+    return await this.usersService.findOne(data)
+  }
+
+  @UseInterceptors(ClassSerializerInterceptor)
+  @Put(':id')
   update(
-    @Param("id") id: number,
+    @Param('id') id: number,
     @Body() updateUserDto: UpdateUserDto
   ): Promise<User> {
     return this.usersService.update(id, updateUserDto);
   }
 
   @UseInterceptors(ClassSerializerInterceptor)
-  @Delete(":id")
-  remove(@Param("id") id: number): Promise<User> {
+  @Delete(':id')
+  remove(@Param('id') id: number): Promise<User> {
     return this.usersService.remove(id);
   }
+
+  @UseInterceptors(ClassSerializerInterceptor)
+  @Post('forgot')
+  async forgotPassword(
+    @Body() forgot: {email: string},
+  ): Promise<string> {
+    const user =  await this.usersService.findOneEmail(forgot.email);
+    const forgotJwt = <string>await this.jwtService.signAsync({id: user.id},
```

Comment:
Please avoid this kind of syntax. More preferable
```suggestion
    const forgotJwt:string = await this.jwtService.signAsync({id: user.id},
```

Reply:
done